File: opcua-webservice-node/src/server/service/transactions/base/UaBrowseNodeTransaction.ts
import { UaBrowseDescription } from "opcua-webapi-ts";
import { NodeManager } from "../../../addressspace/nodemanager";
import { ServiceContext, UaBrowseAdditionalInfo } from "../../../types";
import { UaBrowseTransaction } from "./UaBrowseTransaction";

export class UaBrowseNodeTransaction extends UaBrowseTransaction {
    private readonly _nodeManager: NodeManager;

    constructor(
        serviceContext: ServiceContext,
        nodeToBrowse: UaBrowseDescription,
        additionalInfo: UaBrowseAdditionalInfo,
        handleId: number,
        nodeManager: NodeManager) {
        super(serviceContext, nodeToBrowse, additionalInfo, handleId);
        this._nodeManager = nodeManager;
    }

    async execute(): Promise<void> {
        try {
            const result = this._nodeManager.browse(this.getItem(), this._additionalInfo);
            this._statusCode = result.statusCode;
            this._continuationPoint = result.continuationPoint;
            this._references = result.references;
        } catch (error) {
            this.buildErrorResults(error);
        }
    }
}

File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/base/UaBrowseNodeTransaction.java
package org.opcfoundation.webserver.service.transactions.base;

import org.eclipse.milo.opcua.stack.core.types.structured.BrowseDescription;
import org.eclipse.milo.opcua.stack.core.types.structured.BrowseResult;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManager;
import org.opcfoundation.webserver.types.common.UaBrowseAdditionalInfo;
import org.opcfoundation.webapi.service.types.ServiceContext;

import java.util.Arrays;
import java.util.concurrent.CompletableFuture;

public class UaBrowseNodeTransaction extends UaBrowseTransaction {
    private final NodeManager nodeManager;

    public UaBrowseNodeTransaction(
            ServiceContext serviceContext,
            BrowseDescription nodeToBrowse,
            UaBrowseAdditionalInfo additionalInfo,
            int handleId,
            NodeManager nodeManager)
    {
        super(serviceContext, nodeToBrowse, additionalInfo, handleId);
        this.nodeManager = nodeManager;
    }

    public CompletableFuture<Void> execute()
    {
        return CompletableFuture.supplyAsync(()-> {
            try
            {
                BrowseDescription nodeToBrowse = getItem();
                BrowseResult result = nodeManager.browse(nodeToBrowse, additionalInfo);
                statusCode = result.getStatusCode();
                continuationPoint = result.getContinuationPoint();
                if (null != result.getReferences()) references = Arrays.asList(result.getReferences());
            } catch (Exception e) {
                buildErrorResponse(e);
            }

            return null;
        });
    }
}
